Search version in set

Forgind opened the pull request:
Progress on https://github.com/dotnet/sdk/issues/42367

Current version:

```terminal
C:\Users\User>dotnet workload search version 9.0.100-preview.6.24328.3
Workload manifest ID                               Manifest feature band      Manifest Version
-----------------------------------------------------------------------------------------------------
microsoft.net.workload.emscripten.current          9.0.100-preview.6          9.0.0-preview.6.24327.1
microsoft.net.workload.emscripten.net6             9.0.100-preview.6          9.0.0-preview.6.24327.1
microsoft.net.workload.emscripten.net7             9.0.100-preview.6          9.0.0-preview.6.24327.1
microsoft.net.workload.emscripten.net8             9.0.100-preview.6          9.0.0-preview.6.24327.1
microsoft.net.sdk.android                          9.0.100-preview.6          34.99.0-preview.6.340
microsoft.net.sdk.ios                              9.0.100-preview.6          17.2.9713-net9-p6
microsoft.net.sdk.maccatalyst                      9.0.100-preview.6          17.2.9713-net9-p6
microsoft.net.sdk.macos                            9.0.100-preview.6          14.2.9713-net9-p6
microsoft.net.sdk.maui                             9.0.100-preview.6          9.0.0-preview.6.24327.7
microsoft.net.sdk.tvos                             9.0.100-preview.6          17.2.9713-net9-p6
microsoft.net.workload.mono.toolchain.current      9.0.100-preview.6          9.0.0-preview.6.24327.7
microsoft.net.workload.mono.toolchain.net6         9.0.100-preview.6          9.0.0-preview.6.24327.7
microsoft.net.workload.mono.toolchain.net7         9.0.100-preview.6          9.0.0-preview.6.24327.7
microsoft.net.workload.mono.toolchain.net8         9.0.100-preview.6          9.0.0-preview.6.24327.7
microsoft.net.sdk.aspire                           8.0.100                    8.0.2

C:\Users\User>dotnet workload search version 9.0.100-preview.6.24328.3 --format json
{
  "manifestVersions": {
    "microsoft.net.workload.emscripten.current": "9.0.0-preview.6.24327.1/9.0.100-preview.6",
    "microsoft.net.workload.emscripten.net6": "9.0.0-preview.6.24327.1/9.0.100-preview.6",
    "microsoft.net.workload.emscripten.net7": "9.0.0-preview.6.24327.1/9.0.100-preview.6",
    "microsoft.net.workload.emscripten.net8": "9.0.0-preview.6.24327.1/9.0.100-preview.6",
    "microsoft.net.sdk.android": "34.99.0-preview.6.340/9.0.100-preview.6",
    "microsoft.net.sdk.ios": "17.2.9713-net9-p6/9.0.100-preview.6",
    "microsoft.net.sdk.maccatalyst": "17.2.9713-net9-p6/9.0.100-preview.6",
    "microsoft.net.sdk.macos": "14.2.9713-net9-p6/9.0.100-preview.6",
    "microsoft.net.sdk.maui": "9.0.0-preview.6.24327.7/9.0.100-preview.6",
    "microsoft.net.sdk.tvos": "17.2.9713-net9-p6/9.0.100-preview.6",
    "microsoft.net.workload.mono.toolchain.current": "9.0.0-preview.6.24327.7/9.0.100-preview.6",
    "microsoft.net.workload.mono.toolchain.net6": "9.0.0-preview.6.24327.7/9.0.100-preview.6",
    "microsoft.net.workload.mono.toolchain.net7": "9.0.0-preview.6.24327.7/9.0.100-preview.6",
    "microsoft.net.workload.mono.toolchain.net8": "9.0.0-preview.6.24327.7/9.0.100-preview.6",
    "microsoft.net.sdk.aspire": "8.0.2/8.0.100"
  }
}
```

Forgind committed:
Maybe permit users to find latest set versions

Forgind committed:
Fixes

Forgind committed:
Some improvements

Forgind committed:
Make it not output message

Forgind committed:
Align columns

dsplaisted commented on src/Cli/dotnet/commands/dotnet-workload/search/LocalizableStrings.resx:
```diff
@@ -145,4 +145,10 @@
+    <value>SEARCH_STRING</value>
+  </data>
+  <data name="WorkloadVersionArgumentDescription" xml:space="preserve">
+    <value>Output workload versions associated with the provided workload version.</value>
```
```suggestion
    <value>Output workload manifest versions associated with the provided workload version.</value>
```

dsplaisted commented on src/Cli/dotnet/commands/dotnet-workload/search/WorkloadSearchVersionsCommand.cs:
```diff
@@ -0,0 +1,105 @@
+                if (_workloadSetOutputFormat?.Equals("json", StringComparison.OrdinalIgnoreCase) == true)
+                {
+                    Reporter.WriteLine(JsonSerializer.Serialize(
+                        workloadSet.ManifestVersions.ToDictionary(kvp => kvp.Key.ToString(), kvp => $"{kvp.Value.Version}/{kvp.Value.FeatureBand}")));
```
I think this is (and should be) the same format as the rollback file.  So use `WorkloadSet.ToJson` here instead of duplicating the logic of how it's formatted.

dsplaisted reviewed (commented):
So we've already added `dotnet workload search version` which lists the available workload set versions, right?  It looks like this adds `dotnet workload search version <workloadSetVersion>`, and when the workload set version is specified it displays the manifests versions in that workload set.

I'm not sure about the scenario here.  We do want to make more information available about what is in a workload set, but I think the manifest versions aren't too generally helpful here.  Adding this now with this output format will make it harder to add more useful information (such as release notes, versions of tools such as XCode required, etc) to this command later.

marcpopMSFT commented:
Triage: Summary is that the json format should have slightly more formatting so if we change this to add more data in the future, that is not a breaking change (ie put the manifest info in a subnode).

Forgind committed:
PR feedback

dsplaisted committed:
Various improvements to workload search version command

- Generate error if workload search term and version subcommand are specified
- If no workload sets are found, do not treat as an error, but print message to StdErr
- Fix issue handling preview workload set versions when searching

dsplaisted committed:
Add VM test for workload search version

dsplaisted committed:
Add tests for workload set version mapping

dsplaisted committed:
Add column headers to workload search version <VERSION> table

Also split feature band out into separate column

dsplaisted commented:
I've made some updates to this and would appreciate a quick review.

I added column headers to the table output and split the manifest feature band into a separate column.  I've updated the PR description to show this.  @baronfel, does this look good?

baronfel commented:
Lovely! Thank you for updating the description. Totally agree with making the table output more user-friendly - with the json as the escape hatch we can freely change the user-facing views.

baronfel reviewed (approved):
The new VM tests look very helpful 👍
